core/src/main/java/org/eclipse/hono/util/MessageHelper.java

```diff
@@ -459,39 +440,88 @@ public static Buffer getPayload(final Message msg) {
      * <li>In all other cases, {@code null} is returned.</li>
      * </ul>
      *
-     * @param message The AMQP 1.0 message to parse the body of.
+     * @param msg The AMQP 1.0 message to parse the body of.
      * @return The String representation of the payload data or {@code null} if the message
      *         body does not contain data that can be decoded into a String.
      * @throws NullPointerException if the message is {@code null}.
      */
-    public static String getPayloadAsString(final Message message) {
+    public static String getPayloadAsString(final Message msg) {
+        Objects.requireNonNull(msg);
+        // In case of an AmqpValue containing a String,
+        // we prevent encoding/decoding of the String to/from its UTF-8 bytes.
+        if (msg.getBody() instanceof AmqpValue
+                && ((AmqpValue) msg.getBody()).getValue() instanceof String) {
+            return (String) ((AmqpValue) msg.getBody()).getValue();
+        }
+        return Optional.ofNullable(getPayload(msg)).map(Buffer::toString).orElse(null);
+    }
 
-        Objects.requireNonNull(message);
+    /**
+     * Gets the size of the payload data contained in a message's body.
+     * <p>
+     * If there is no body section or if its type is unsupported, {@code 0} is returned.
+     *
+     * @param msg The AMQP 1.0 message to parse the body of.
+     * @return The payload size in bytes, 0 if message body is {@code null} or unsupported.
+     * @throws NullPointerException if the message is {@code null}.
+     */
+    public static int getPayloadSize(final Message msg) {
+        Objects.requireNonNull(msg);
+        return Optional.ofNullable(getPayloadByteArray(msg)).map(bytes -> bytes.length).orElse(0);
+    }
 
-        if (message.getBody() == null) {
-            LOG.trace("message has no body");
-            return null;
-        }
+    /**
+     * Checks whether the payload data contained in a message's body is returned as {@code null} by
+     * {@link #getPayload(Message)} or {@link #getPayloadAsString(Message)} and, if that is the case,
+     * returns the reason for that.
+     *
+     * @param msg The AMQP 1.0 message to parse.
+     * @return A reason string or {@code null} if the payload is not {@code null}.
+     * @throws NullPointerException if the message is {@code null}.
+     */
+    public static String getReasonIfPayloadNull(final Message msg) {
```

Comment:
This looks a little odd to me. How about renaming this to something like *assertSupportedPayloadType* and throw an exception holding a description in case of an invalid payload?

Reply:
Throwing an exception in that message doesn't fit the use case of the `Command` class and just adds clutter there.
I've renamed the method to `getUnsupportedPayloadReason` now (returning an Optional).
When the need for a `assertSupportedPayloadTyp` method arises, we can add that, simply invoking `getUnsupportedPayloadReason` internally.

Reply:
Is this method being used anywhere else? If this is a very Command specific requirement, then we could also simply embed this method into the Command class. WDYT?

Reply:
Ok, I've moved the method to the `Command` class.
I've also updated the new `org.eclipse.hono.adapter.client.command.Command` class.